ql/src/java/org/apache/hadoop/hive/ql/parse/repl/dump/io/FileOperations.java

```diff
@@ -136,18 +137,41 @@ private void copyMmPath() throws LoginException, IOException {
     }
   }
 
-
-
   /**
    * This needs the root data directory to which the data needs to be exported to.
    * The data export here is a list of files either in table/partition that are written to the _files
    * in the exportRootDataDir provided.
    */
   private void exportFilesAsList() throws SemanticException, IOException {
-    // This is only called for replication that handles MM tables; no need for mmCtx.
-    try (BufferedWriter writer = writer()) {
-      for (Path dataPath : dataPathList) {
-        writeFilesList(listFilesInDir(dataPath), writer, AcidUtils.getAcidSubDir(dataPath));
+    if (dataPathList.isEmpty()) {
+      return;
+    }
+    boolean done = false;
+    int repeat = 0;
+    while (!done) {
+      // This is only called for replication that handles MM tables; no need for mmCtx.
+      try (BufferedWriter writer = writer()) {
+        for (Path dataPath : dataPathList) {
+          writeFilesList(listFilesInDir(dataPath), writer, AcidUtils.getAcidSubDir(dataPath));
+        }
+        done = true;
+      } catch (IOException e) {
+        repeat++;
+        logger.info("writeFilesList failed with exception : " + e.getMessage() + " for " + repeat + " time");
+        if (repeat >= MAX_IO_ERROR_RETRY) {
+          logger.error("exporting data files in dir : " + dataPathList + " to " + exportRootDataDir + " failed");
+          throw e;
+        }
+        try {
+          logger.info(" sleep for 100 milliseconds before retry " + (repeat));
+          Thread.sleep(100);
+        } catch (InterruptedException timerEx) {
+          logger.info("thread sleep interrupted", timerEx.getMessage());
+        }
+        // in case of io error, reset the file system object
+        dataFileSystem = dataPathList.get(0).getFileSystem(hiveConf);
+        exportFileSystem = exportRootDataDir.getFileSystem(hiveConf);
+        exportFileSystem.deleteOnExit(new Path(exportRootDataDir, EximUtil.FILES_NAME));
```

Comment:
What is exportFileSystem.deleteOnExit for?

Reply:
there is a check for directory exist ..it will fail in failure ..so we should delete it first